http-client/Network/HTTP/Client/Types.hs

```diff
@@ -634,6 +634,11 @@ data Request = Request
     -- ^ List of header values being redacted in case we show Request.
     --
     -- @since 0.7.13
+
+    , earlyHintHeaderReceived :: Header -> IO ()
```

Comment:
This signature surprises me a bit, I'd have thought we'd send all the headers at once. There shouldn't be a significant time delay in slurping in all the headers, and from the user's perspective there's no way to distinguish if all early hint headers have been received or not given the single-header API.

Reply:
I was thinking it's theoretically possible for a web server to be doing some streaming server-side rendering, and outputting "Link" headers one by one as it encounters them. So then there might be time delays between the different headers.

However, it seems like the current [Node.js](https://nodejs.org/api/http.html#responsewriteearlyhintshints-callback) implementation sends them as a batch.

You're right, the user does need to know when they've all been received. I'd suggest one of the following:
a) `Maybe Header -> IO ()` (sends a `Just` on every header, then `Nothing` to indicate completion)
b) `[Header] -> IO ()` (sends them all as a batch)

Option b) was my original plan and would probably be fine, and would make it easier to write a reverse proxy. It would be messy to change our minds later though.

The [RFC](https://www.rfc-editor.org/rfc/rfc8297#section-2) says the following, which suggests to me that a server which wishes to do streaming rendering can just send multiple early hint batches.

```
A server MAY emit multiple 103 (Early Hints) responses with
additional header fields as new information becomes available while
the request is being processed.
```

Reply:
I think (b) is fine, I'm not aware of any other API where individual headers can be streamed, so most likely the multiple 103s approach would be used if that needs to work reliably.